packages/postgres-database/tests/projects/test_projects_version_control.py

```diff
@@ -0,0 +1,285 @@
+
+
+class ReposOrm(BaseOrm[int]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_repos,
+            connection,
+            readonly={"id", "created", "modified"},
+        )
+
+
+class BranchesOrm(BaseOrm[int]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_branches,
+            connection,
+            readonly={"id", "created", "modified"},
+        )
+
+
+class CommitsOrm(BaseOrm[int]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_commits,
+            connection,
+            readonly={"id", "created", "modified"},
+            # pylint: disable=no-member
+            writeonce=set(c for c in projects_vc_commits.columns.keys()),
+        )
+
+
+class TagsOrm(BaseOrm[int]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_tags,
+            connection,
+            readonly={"id", "created", "modified"},
+        )
+
+
+class ProjectsOrm(BaseOrm[str]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects,
+            connection,
+            readonly={"id", "creation_date", "last_change_date"},
+            writeonce={"uuid"},
+        )
+
+
+class SnapshotsOrm(BaseOrm[str]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_snapshots,
+            connection,
+            writeonce={"checksum"},  # TODO:  all? cannot delete snapshots?
+        )
+
+
+class HeadsOrm(BaseOrm[int]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_heads,
+            connection,
+            writeonce={"repo_id"},
+        )
+
+
+# -------------
+
+
+def eval_checksum(workbench: Dict[str, Any]):
+    # FIXME: prototype
+    block_string = json.dumps(workbench, sort_keys=True).encode("utf-8")
+    raw_hash = hashlib.sha256(block_string)
+    return raw_hash.hexdigest()
+
+
+def eval_snapshot_uuid(repo: RowProxy, commit: RowProxy) -> UUID:
+    assert repo.id == commit.repo_id  # nosec
+    return uuid3(UUID(repo.project_uuid), f"{repo.id}.{commit.snapshot_checksum}")
+
+
+async def add_snapshot(
+    project_wc: RowProxy, checksum: str, repo: RowProxy, conn: SAConnection
+) -> str:
+    snapshot_orm = SnapshotsOrm(conn)
+    snapshot_checksum = checksum
+    row_id = await snapshot_orm.insert(
+        checksum=checksum,
+        content={"workbench": project_wc.workbench, "ui": project_wc.ui},
+    )
+    assert row_id == checksum
+    return checksum
+
+
+async def test_basic_workflow(project: RowProxy, conn: SAConnection):
+
+    # git init
+    async with conn.begin():
+        # create repo
+        repo_orm = ReposOrm(conn)
+        repo_id = await repo_orm.insert(project_uuid=project.uuid)
+        assert repo_id is not None
+
+        repo_orm.pin(rowid=repo_id)
```

Comment:
actually I am a bit puzzled by that *pin* thing. is that an equivalent to doing *select* and then *fetch*? 
minor: why not just doing fetch(rowid=repo_id)?

Reply:
you can also do that ...  (look at the arguments of ``fetch``)

the ``pin`` is to avoid having to pass the ``rowid`` all the time. As a convenience ... but you do not need to use it

Reply:
I renamed it as ``set_defaults`` ... which shows better the intention.
I found it very handy but it is true that keeping a state like that the class is a bit of magic.
I am open to remove it but let's give it a chance :-)